### PR TITLE
Add support for defining default policies

### PR DESCRIPTION
**Pull request by slluis**
This patch introduces a new Default policy set, which defines the default
policies when there are no solution-defined or user-defined policies.

Before this patch, default policies were obtained by just creating an intance
of the policy type. This is still the case when a policy is not set for the
Default policy set.

This new Default policy set is mostly useful for defining default policies
for specific file types (which was impossible to do).

### CODE REVIEW

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Diagnostics/InconsistentNaming/NameConventionRule.cs [hunk 1]
```diff
@@ -126,6 +126,21 @@ public NameConventionRule Clone ()
 			};
 		}
 
+		public bool Equals (NameConventionRule other)
+		{
+			var empty = new string [0];
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Diagnostics/InconsistentNaming/NameConventionRule.cs]: `Array.Empty<string> ()`?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.Policies/PolicySet.cs [hunk 2]
```diff
@@ -78,11 +82,16 @@ public PolicySet ()
 		
 		protected override T GetDefaultPolicy<T> ()
 		{
+			// The default policy set always resturns a value for any type of policy.
+			if (IsDefaultSet)
+				return new T ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects.Policies/PolicySet.cs]: How does this compile without a `new()` constraint on `T`?

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Diagnostics/InconsistentNaming/NameConventionRule.cs (hunk 1 above)
Comment (slluis) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Diagnostics/InconsistentNaming/NameConventionRule.cs]: Not available in .NET 4.5.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.Policies/PolicySet.cs (hunk 2 above)
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects.Policies/PolicySet.cs]: The constraint is specified in the virtual method that this overrides.

Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects.Policies/PolicySet.cs]: Missed the override part, 👍

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Diagnostics/InconsistentNaming/NameConventionRule.cs (hunk 1 above)
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Diagnostics/InconsistentNaming/NameConventionRule.cs]: *patiently waiting for roslyn-ivt*

**Pull request merged**